File: UsingFMDB-Objective-C/UsingFMDB-Objective-C/BooksViewController.m
//
//  ViewController.m
//  UsingFMDB-Objective-C
//
//  Created by akabeko on 2016/11/21.
//  Copyright © 2016年 akabeko. All rights reserved.
//

#import "BooksViewController.h"
#import "EditBookViewController.h"
#import "AppDelegate.h"
#import "AppStore.h"
#import "BookStore.h"
#import "Book.h"

/** Segur fot the edit book. */
static NSString * const kSegueEditBook = @"EditBook";

////////////////////////////////////////////////////////////////////////////////
@interface BooksViewController () <EditBookViewDelegate>

/** A value indicating that a new book should be created. */
@property (nonatomic) BOOL creatingBook;

@end

////////////////////////////////////////////////////////////////////////////////
@implementation BooksViewController

#pragma mark - UIViewController

/**
 * Called after the controller's view is loaded into memory.
 *
 * @see https://developer.apple.com/reference/uikit/uiviewcontroller/1621495-viewdidload?language=objc
 */
- (void)viewDidLoad {
    [super viewDidLoad];

    self.title = @"Books";

    UIBarButtonItem *button = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAdd target:self action:@selector(didTouchCreateBookButton:)];
    self.navigationItem.leftBarButtonItem = button;
    
    self.navigationItem.rightBarButtonItem = self.editButtonItem;
}

/**
 * Notifies the view controller that its view is about to be removed from a view hierarchy.
 *
 * @param animated If YES, the disappearance of the view is being animated.
 *
 * https://developer.apple.com/reference/uikit/uiviewcontroller/1621485-viewwilldisappear?language=objc
 */
- (void)viewWillDisappear:(BOOL)animated {
    [self setEditing:NO];

    [super viewWillDisappear:animated];
}

/**
 * Sent to the view controller when the app receives a memory warning.
 *
 * @see https://developer.apple.com/reference/uikit/uiviewcontroller/1621409-didreceivememorywarning?language=objc
 */
- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

/**
 * Notifies the view controller that a segue is about to be performed.
 *
 * @param segue  The segue object containing information about the view controllers involved in the segue.
 * @param sender The object that initiated the segue. You might use this parameter to perform different actions based on which control (or other object) initiated the segue.
 *
 * @see https://developer.apple.com/reference/uikit/uiviewcontroller/1621490-prepareforsegue
 */
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if( [[segue identifier] isEqualToString:kSegueEditBook] ) {
        EditBookViewController *vc = [segue destinationViewController];
        vc.originalBook = self.creatingBook ? nil : [self bookAtIndexPath:self.tableView.indexPathForSelectedRow];
        vc.delegate     = self;
    }
}

#pragma mark - Action

/**
 * Occurs when the book creation button is touched.
 *
 * @param sender Target of the event.
 */
- (void)didTouchCreateBookButton:(id)sender {
    self.creatingBook = YES;
    [self performSegueWithIdentifier:kSegueEditBook sender:self];
}

#pragma mark - UITableViewDataSource

/**
 * Asks the data source to return the number of sections in the table view.
 *
 * @param tableView An object representing the table view requesting this information.
 *
 * @return The number of sections in tableView. The default value is 1.
 *
 * @see https://developer.apple.com/reference/uikit/uitableviewdatasource/1614860-numberofsectionsintableview?language=objc
 */
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    BookStore *store = [self bookStore];
    return [store.authors count];
}

/**
 * Tells the data source to return the number of rows in a given section of a table view.
 *
 * @param tableView The table-view object requesting this information.
 * @param section   An index number identifying a section in tableView.
 *
 * @return The number of rows in section.
 *
 * @see https://developer.apple.com/reference/uikit/uitableviewdatasource/1614931-tableview?language=objc
 */
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    BookStore *store  = [self bookStore];
    NSString  *author = [store.authors objectAtIndex:section];
    NSArray   *books  = [store.booksByAuthor objectForKey:author];

    return [books count];
}

/**
 * Asks the data source for the title of the header of the specified section of the table view.
 *
 * @param tableView The table-view object asking for the title.
 * @param section   An index number identifying a section of tableView.
 *
 * @return A string to use as the title of the section header. If you return nil , the section will have no title.
 *
 * @see https://developer.apple.com/reference/uikit/uitableviewdatasource/1614850-tableview?language=objc
 */
- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    BookStore *store = [self bookStore];
    return [store.authors objectAtIndex:section];
}

/**
 * Asks the data source for a cell to insert in a particular location of the table view.
 *
 * @param tableView A table-view object requesting the cell.
 * @param indexPath An index path locating a row in tableView.
 *
 * @return An object inheriting from UITableViewCell that the table view can use for the specified row. An assertion is raised if you return nil.
 *
 * @see https://developer.apple.com/reference/uikit/uitableviewdatasource/1614861-tableview?language=objc
 */
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *CellIdentifier = @"Cell";
    UITableViewCell *cell           = [tableView dequeueReusableCellWithIdentifier:CellIdentifier forIndexPath:indexPath];
    UILabel         *label          = [cell viewWithTag:1];
    Book            *book           = [self bookAtIndexPath:indexPath];

    label.text = book.title;

    return cell;
}

/**
 * Asks the data source to commit the insertion or deletion of a specified row in the receiver.
 *
 * @param tableView    The table-view object requesting the insertion or deletion.
 * @param editingStyle The cell editing style corresponding to a insertion or deletion requested for the row specified by indexPath. Possible editing styles are UITableViewCellEditingStyleInsert or UITableViewCellEditingStyleDelete.
 * @param indexPath    An index path locating the row in tableView.
 *
 * @see https://developer.apple.com/reference/uikit/uitableviewdatasource/1614871-tableview?language=objc
 */
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        BookStore *store = [self bookStore];
        Book      *book  = [self bookAtIndexPath:indexPath];

        if ([store remove:book]) {
            [self.tableView reloadData];
        }
    }
}

#pragma mark - UITableViewDelegate

/**
 * Tells the delegate that the specified row is now selected.
 *
 * @param tableView A table-view object informing the delegate about the new row selection.
 * @param indexPath An index path locating the new selected row in tableView.
 *
 * @see https://developer.apple.com/reference/uikit/uitableviewdelegate/1614877-tableview?language=objc
 */
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    self.creatingBook = NO;
    [self performSegueWithIdentifier:kSegueEditBook sender:self];
}

#pragma mark - EditBookViewDelegate

/**
 * Occurs when editing or creation of a book is completed.
 *
 * @param oldBook Old book data.
 * @param newBook New book data.
 */
- (void)didFinishEditBook:(Book *)oldBook newBook:(Book *)newBook{
    BookStore *store    = [self bookStore];
    BOOL      succeeded = NO;

    if (newBook.bookId == kBookIdNone) {
        succeeded = [store add:newBook];
    } else {
        succeeded = [store update:oldBook newBook:newBook];
    }

    if (succeeded) {
        [self.tableView reloadData];
    }
}

#pragma mark - Private

/**
 * Get the book at index path.
 *
 * @param indexPath An index path locating a row in tableView.
 *
 * @return Book data.
 */
- (Book *)bookAtIndexPath:(NSIndexPath *)indexPath {
    BookStore *store  = [self bookStore];
    NSString  *author = [store.authors objectAtIndex:indexPath.section];
    NSArray   *books  = [store.booksByAuthor objectForKey:author];

    return [books objectAtIndex:indexPath.row];
}

/**
 * Get the book sore.
 *
 * @return Instance of the book store.
 */
- (BookStore *)bookStore {
    AppDelegate *app   = (AppDelegate *)[[UIApplication sharedApplication] delegate];
    return app.appStore.bookStore;
}

@end

File: UsingFMDB-Swift/UsingFMDB-Swift/BooksViewController.swift
//
//  BooksViewController.swift
//  UsingFMDB-Swift
//
//  Created by akabeko on 2016/12/17.
//  Copyright © 2016年 akabeko. All rights reserved.
//

import UIKit

/// List view for a books.
class BooksViewController: UITableViewController, EditBookViewControllerDelegate {
    /// Segue fot the edit book.
    private static let SegueEditBook = "EditBook"

    /// Name for the cell.
    private static let CellIdentifier = "Cell"

    /// A value indicating that a new book should be created.
    private var creatingBook = false

    /// Called after the controller's view is loaded into memory.
    override func viewDidLoad() {
        super.viewDidLoad()

        self.title = "Books"

        let button = UIBarButtonItem(barButtonSystemItem: UIBarButtonSystemItem.add, target: self, action: #selector(didTouchCreateBookButton(sender:)))
        self.navigationItem.leftBarButtonItem  = button
        self.navigationItem.rightBarButtonItem = self.editButtonItem
    }

    /// Notifies the view controller that its view is about to be removed from a view hierarchy.
    ///
    /// - Parameter animated: If true, the disappearance of the view is being animated.
    override func viewWillAppear(_ animated: Bool) {
        self.setEditing(false, animated: animated)
        super.viewWillAppear(animated)
    }

    /// Sent to the view controller when the app receives a memory warning.
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }

    /// Notifies the view controller that a segue is about to be performed.
    ///
    /// - Parameters:
    ///   - segue: The segue object containing information about the view controllers involved in the segue.
    ///   - sender: The object that initiated the segue. You might use this parameter to perform different actions based on which control (or other object) initiated the segue.
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == BooksViewController.SegueEditBook {
            let vc  = segue.destination as! EditBookViewController
            vc.originalBook = self.creatingBook ? nil : self.bookAtIndexPath(indexPath: self.tableView.indexPathForSelectedRow!)
            vc.deletate     = self
        }
    }

    /// Occurs when editing or creation of a book is completed.
    ///
    /// - Parameters:
    ///   - viewController: Sender.
    ///   - oldBook:        Old book data.
    ///   - newBook:        New book data.
    func didFinishEditBook(viewController: EditBookViewController, oldBook: Book?, newBook: Book) {
        let store   = self.bookStore()
        var success = false;
        if (newBook.bookId == Book.BookIdNone) {
            success = store.add(book: newBook)
        } else {
            success = store.update(oldBook: oldBook!, newBook: newBook)
        }
  
        if success {
            self.tableView.reloadData()
        }
    }

    /// Occurs when the book creation button is touched.
    ///
    /// - Parameter sender: Target of the event.
    func didTouchCreateBookButton(sender: Any?) {
        self.creatingBook = true
        self.performSegue(withIdentifier: BooksViewController.SegueEditBook, sender: self)
    }

    /// Asks the data source to return the number of sections in the table view.
    ///
    /// - Parameter tableView: An object representing the table view requesting this information.
    /// - Returns: The number of sections in tableView. The default value is 1.
    override func numberOfSections(in tableView: UITableView) -> Int {
        let store  = self.bookStore()
        return store.authors.count
    }

    /// Tells the data source to return the number of rows in a given section of a table view.
    ///
    /// - Parameters:
    ///   - tableView: The table-view object requesting this information.
    ///   - section:   An index number identifying a section in tableView.
    /// - Returns: The number of rows in section.
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let store  = self.bookStore()
        let author = store.authors[section]
        let books  = store.booksByAuthor[author]

        return (books?.count)!
    }

    /// Asks the data source for the title of the header of the specified section of the table view.
    ///
    /// - Parameters:
    ///   - tableView: The table-view object asking for the title.
    ///   - section:   An index number identifying a section of tableView.
    /// - Returns: A string to use as the title of the section header. If you return nil , the section will have no title.
    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        let store = self.bookStore()
        return store.authors[section]
    }

    /// Asks the data source for a cell to insert in a particular location of the table view.
    ///
    /// - Parameters:
    ///   - tableView: A table-view object requesting the cell.
    ///   - indexPath: An index path locating a row in tableView.
    /// - Returns: An object inheriting from UITableViewCell that the table view can use for the specified row. An assertion is raised if you return nil.
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell  = tableView.dequeueReusableCell(withIdentifier: BooksViewController.CellIdentifier, for: indexPath)
        let label = cell.viewWithTag(1) as! UILabel
        let book  = self.bookAtIndexPath(indexPath: indexPath)

        label.text = book.title

        return cell
    }

    /// Asks the data source to commit the insertion or deletion of a specified row in the receiver.
    ///
    /// - Parameters:
    ///   - tableView:    The table-view object requesting the insertion or deletion.
    ///   - editingStyle: The cell editing style corresponding to a insertion or deletion requested for the row specified by indexPath. Possible editing styles are UITableViewCellEditingStyleInsert or UITableViewCellEditingStyleDelete.
    ///   - indexPath:    An index path locating the row in tableView.
    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle == UITableViewCellEditingStyle.delete {
            let store = self.bookStore()
            let book  = self.bookAtIndexPath(indexPath: indexPath)
            if store.remove(book: book) {
                self.tableView.reloadData()
            }
        }
    }

    /// Tells the delegate that the specified row is now selected.
    ///
    /// - Parameters:
    ///   - tableView: A table-view object informing the delegate about the new row selection.
    ///   - indexPath: An index path locating the new selected row in tableView.
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        self.creatingBook = false
        self.performSegue(withIdentifier: BooksViewController.SegueEditBook, sender: self)
    }

    /// Get the book at index path.
    ///
    /// - Parameter indexPath: An index path locating a row in tableView.
    /// - Returns: Book data.
    private func bookAtIndexPath(indexPath: IndexPath) -> Book {
        let store  = self.bookStore()
        let author = store.authors[indexPath.section]
        let books  = store.booksByAuthor[author]

        return books![indexPath.row]
    }

    /// Get the book sore.
    ///
    /// - Returns: Instance of the book store.
    func bookStore() -> BookStore {
        let app = UIApplication.shared.delegate as! AppDelegate
        return app.appStore.bookStore
    }
}

